src/runtime/tuple_parser.c

```diff
@@ -52,49 +49,59 @@ static parser combinate(heap h, selfparser p)
     return (parser)closure(h, selfinate, h, p);
 }
 
-
 closure_function(1, 2, parser, til_newline,
                  parser, finish,
                  parser, self, character, in)
 {
-    if (in == '\n') return bound(finish);
+    if (in == '\n') {
+        parser f = bound(finish);
+        deallocate_closure(self);
+        closure_finish();
+        return f;
+    }
     return self;
 }
 
+// leaky; no finite lifespan
 closure_function(2, 2, parser, eat_whitespace,
                  heap, h, parser, finish,
                  parser, self, character, in)
 {
-    if (in == '#') {
+    if (in == '#')
         return combinate(bound(h), closure(bound(h), til_newline, self));
-    }
 
-    if (member(whitespace, in)) return self;
+    if (member(whitespace, in))
+        return self;
+
     return apply(bound(finish), in);
```

Comment:
doesn't this conditional leg mean that this closure is done and can be deallocated?

Reply:
Yes, should go away as above...same for below